ansible_collections/arista/avd/roles/eos_designs/python_modules/ip_addressing/__init__.py

```diff
@@ -0,0 +1,346 @@
+
+from collections import ChainMap
+from functools import cached_property
+import ipaddress
+import importlib
+from ansible_collections.arista.avd.plugins.plugin_utils.avdfacts import AvdFacts
+from ansible_collections.arista.avd.plugins.plugin_utils.utils import AristaAvdError, get
+
+
+DEFAULT_AVD_IP_ADDRESSING_PYTHON_MODULE = "ansible_collections.arista.avd.roles.eos_designs.python_modules.ip_addressing"
+DEFAULT_AVD_IP_ADDRESSING_PYTHON_CLASS_NAME = "AvdIpAddressing"
+
+
+class AvdIpAddressing(AvdFacts):
+
+    '''
+    Class used to render IP addresses either from custom Jinja2 templates or using default Python Logic
+
+    Since some templates might contain certain legacy variables (switch_*),
+    those are mapped from the switch.* model
+
+    This class is imported adhoc based on the variable `templates.ip_addressing.python_module` so it can
+    be overridden by a custom python class.
+    '''
+
+    def _ip(self, pool: str, prefixlen: int, subnet_offset: int, ip_offset: int) -> str:
+        network = ipaddress.ip_network(pool, strict=False)
+        subnets = network.subnets(new_prefix=prefixlen)
+        if isinstance(subnets, list):
+            # It seems that some versions of python return a list instead of the documented iterator
+            subnets = iter(subnets)
+        subnet_index = 0
+        try:
+            while subnet_index <= subnet_offset:
+                subnet = next(subnets)
+                subnet_index += 1
+        except StopIteration as e:
+            raise AristaAvdError(f"Unable to get {subnet_offset+1} /{prefixlen} subnets from pool {pool}") from e
+        ips = subnet.hosts()
+        if isinstance(ips, list):
+            # It seems that some versions of python return a list instead of the documented iterator
+            ips = iter(ips)
+        ip_index = 0
+        try:
+            while ip_index <= ip_offset:
+                ip = next(ips)
+                ip_index += 1
+        except StopIteration as e:
+            raise AristaAvdError(f"Unable to get {ip_offset+1} hosts in subnet {subnet} taken from pool {pool}") from e
```

Comment:
it looks like math on IP works (python 3.10) so could this maybe be made simple?
```
>>> ipaddress.IPv4Address("42.42.42.42")
IPv4Address('42.42.42.42')
>>> ii = _
>>> ii+1
IPv4Address('42.42.42.43')
>>> ii+2
IPv4Address('42.42.42.44')
>>> ii+3
IPv4Address('42.42.42.45')
>>> ii+4
IPv4Address('42.42.42.46')
```

Like checking if ip_offset is longer than subnet length maybe? - this while loop seem unecessary except if I am missing something?

Reply:
Let me experiment a bit. But this iterator seems quite effective.

Reply:
I tried refactoring the code to this 
```python 
    def _ip(self, pool: str, prefixlen: int, subnet_offset: int, ip_offset: int) -> str:
        network = ipaddress.ip_network(pool, strict=False)
        prefixlen_diff = prefixlen - network.prefixlen
        subnets_in_pool = 2**prefixlen_diff
        if subnet_offset >= subnets_in_pool:
            raise AristaAvdError(
                f"Unable to get {subnet_offset+1} /{prefixlen} subnets from pool {pool}"
            )

        if network.version == 4:
            subnetsize = 2 ** (32 - prefixlen)
        else:
            subnetsize = 2 ** (128 - prefixlen)

        subnet = ipaddress.ip_network(
            (int(network.network_address) + subnet_offset * subnetsize, prefixlen)
        )
        if ip_offset >= subnet.num_addresses:
            raise AristaAvdError(
                f"Unable to get {ip_offset+1} hosts in subnet {subnet} taken from pool {pool}"
            )
        return subnet[ip_offset]
```
While it makes it simpler to read, it is also slightly slower for the very large inventory I am testing with. I would also expect it to be much slower for ipv6, since it calculates the large mask sizes. We could add a prefixlen indexed list with precomputed values. It would only be 128 items and a single lookup instead of calculating 2**128 etc.

Reply:
Another attempt which seems to have ~ the same performance as the iterator:
```python
    def _ip(self, pool: str, prefixlen: int, subnet_offset: int, ip_offset: int) -> str:
        pool_network = ipaddress.ip_network(pool, strict=False)
        prefixlen_diff = prefixlen - pool_network.prefixlen
        subnet_size = (int(pool_network.hostmask) + 1) >> prefixlen_diff
        if (subnet_offset + 1) * subnet_size > pool_network.num_addresses:
            raise AristaAvdError(
                f"Unable to get {subnet_offset + 1} /{prefixlen} subnets from pool {pool}"
            )
        subnet = ipaddress.ip_network(
            (int(pool_network.network_address) + subnet_offset * subnet_size, prefixlen)
        )
        try:
            ip = subnet[ip_offset]
        except IndexError as e:
            raise AristaAvdError(
                f"Unable to get {ip_offset+1} hosts in subnet {subnet} taken from pool {pool}"
            ) from e
        return str(ip)
```